include/divergence_block_chain.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include "communicator.h"
+#include "recorders.h"
+#include <mutex>
+#include <thread>
+#include <vector>
+
+namespace celerity::detail {
+/**
+ *  @brief This class is a wrapper around a 1D vector that allows us to access it as a 2D array.
+ *
+ *  It is used to send the task hashes to other nodes using MPI while keeping the code simple and readable.
+ */
+template <typename T>
+struct vector_2d {
+  public:
+	vector_2d(size_t width, size_t height) : m_data(width * height), m_width(width){};
+
+	const T& operator[](id<2> ij) const {
+		assert(ij[0] * m_width + ij[1] < m_data.size());
+		return m_data[ij[0] * m_width + ij[1]];
+	}
+
+	T* data() { return m_data.data(); }
+
+  private:
+	std::vector<T> m_data;
+	const size_t m_width;
+};
+
+/**
+ *  @brief This class gives a view into a const vector.
+ *
+ *	It is used to give us the currently unhashed task records while keeping track of the offset and width.
+ */
+template <typename T>
+struct sliding_window {
+  public:
+	sliding_window(const std::vector<T>& value) : m_value(value) {}
+
+	const T& operator[](size_t i) const {
+		assert(i >= 0 && i < m_width);
+		return m_value[m_offset + i];
+	}
+
+	size_t size() {
+		m_width = m_value.size() - m_offset;
+		return m_width;
+	}
+
+	void slide(size_t i) {
+		assert(i == 0 || (i >= 0 && i <= m_width));
+		m_offset += i;
+		m_width -= i;
+	}
+
+  private:
+	const std::vector<T>& m_value;
+	size_t m_offset = 0;
+	size_t m_width = 0;
+};
+
+using task_hash = size_t;
+using task_hash_data = vector_2d<task_hash>;
+using divergence_map = std::unordered_map<task_hash, std::vector<node_id>>;
+
+/**
+ *  @brief This class checks for divergences of tasks between nodes.
+ *
+ *	It is responsible for collecting the task hashes from all nodes and checking for differences -> divergence.
+ *	When a divergence is found, the task record for the diverging task is printed and the program is terminated.
+ *  Additionally it also checks for deadlocks and prints a warning if one is detected.
+ */
+
+class divergence_block_chain {
+	friend struct divergence_block_chain_testspy;
+
+  public:
+	divergence_block_chain(const std::vector<task_record>& task_recorder, std::unique_ptr<communicator> comm)
```

Comment:
```suggestion
	divergence_block_chain(const std::vector<task_record>* task_records, std::unique_ptr<communicator> comm)
```

Unfortunately using a const reference here means that it could also accept a lifetime-extended temporary; consider using a pointer instead (suggested by @fknorr!).